lib/memdebug.man
.\" RCSID @(#)$Id: memdebug.man,v 1.1 2002/08/27 21:07:41 rkowen Exp $
.\" LIBDIR
.TH "MEMDEBUG" "3rko" "12 May 1995"
.SH NAME
memdebug \- a set of macros that provides a front-end for the memory allocation
routines to help find memory leaks.  Supports only the ANSI-C routines:
.IR calloc , free , malloc , realloc.

.SH SYNOPSIS

 \.\.\. usual standard headers
 #define MEMDEBUG
 #include "memdebug.h"
 \.\.\. code that calls the memory allocation routines

.SH DESCRIPTION
.I memdebug
provides several front-end routines for the ANSI-C memory allocation
routines.  It outputs the routine name, pointer(s), file name,
and line number to output.  This can be analyzed to see that allocations
are matched with deallocations.

.P
The only user callable function is
.I FILE *memdebug_output(FILE *output)
which will set the output stream for the memdebug routines
if output is not NULL, else it will return the FILE *stream
that will be used if output is set NULL.  The default stream
is
.IR stderr .
This routine should be called before any memory allocations are done.
.P
This routine is provide, because it may be more convenient to set it to
.I stdout
or whatever file stream the regular output goes to
so that the MEMDEBUG lines can be
synchronized between the program and the memory
allocation calls.

.SH EXAMPLE
.nf
#include <stdlib.h>
#include <stdio.h>

#define MEMDEBUG
#include "memdebug.h"		/* include this AFTER stdlib.h */

typedef struct Trial {
	int a;
	double b;
	char c[5];
	struct Trial *next;
} trial;

int main() {

	size_t num = 20;
	int *ia = (int *) NULL;
	double *da = (double *) NULL;
	char *ca = (char *) NULL;
	trial *ta = (trial *) NULL;

	printf("Some regular output\n");
	memdebug_output(stdout);	/* set memdebug output */

	if ((ia = (int *) malloc(num * sizeof(int))) == NULL)
		(void) fprintf(stderr,"malloc error\n");
	free(ia);
	ia = (int *) NULL;

	if ((ia = (int *) malloc(2 * num * sizeof(int))) == NULL)
		(void) fprintf(stderr,"malloc error\n");
	if ((ia = (int *) realloc(ia, 4 * num * sizeof(int))) == NULL)
		(void) fprintf(stderr,"realloc error\n");

	if ((da = (double *) malloc(num * sizeof(double))) == NULL)
		(void) fprintf(stderr,"malloc error\n");

	if ((ta = (trial *) calloc(num, sizeof(trial))) == NULL)
		(void) fprintf(stderr,"malloc error\n");
	free(ta);
	ta = (trial *) NULL;

	if ((ta = (trial *) calloc(2 * num, sizeof(trial))) == NULL)
		(void) fprintf(stderr,"malloc error\n");

	if ((da = (double *) realloc(da, 2 * num * sizeof(double))) == NULL)
		(void) fprintf(stderr,"realloc error\n");
	free(ia);
	ia = (int *) NULL;

	if ((ca = (char *) malloc(num * sizeof(char))) == NULL)
		(void) fprintf(stderr,"malloc error\n");
	printf("Some more regular output\n");

	free(da);
	free(ta);
	free(ca);
	free(ia);

	return 0;
}
.fi
.P
The output generated by the above code should look something like:
.nf
Some regular output
RKOMEM:         malloc : tmemdebug.c 36 : 0x8049fd8 80
RKOMEM:         free   : tmemdebug.c 38 : 0x8049fd8
RKOMEM:         malloc : tmemdebug.c 41 : 0x8049fd8 160
RKOMEM: realloc_free   : tmemdebug.c 43 : 0x8049fd8
RKOMEM: realloc_malloc : tmemdebug.c 43 : 0x8049fd8 320
RKOMEM:         malloc : tmemdebug.c 46 : 0x804a120 160
RKOMEM:         calloc : tmemdebug.c 49 : 0x804a1c8 20 24
RKOMEM:         free   : tmemdebug.c 51 : 0x804a1c8
RKOMEM:         calloc : tmemdebug.c 54 : 0x804a1c8 40 24
RKOMEM: realloc_free   : tmemdebug.c 57 : 0x804a120
RKOMEM: realloc_malloc : tmemdebug.c 57 : 0x804a590 320
RKOMEM:         free   : tmemdebug.c 59 : 0x8049fd8
RKOMEM:         malloc : tmemdebug.c 62 : 0x8049fd8 20
Some more regular output
RKOMEM:         free   : tmemdebug.c 67 : 0x804a590
RKOMEM:         free   : tmemdebug.c 68 : 0x804a1c8
RKOMEM:         free   : tmemdebug.c 69 : 0x8049fd8
RKOMEM:         free   : tmemdebug.c 70 : (null)
.fi

.SH SEE ALSO
.\" crontab(1),stat(2),libmon(8)
memdbg(1rko), calloc(3), free(3), malloc(3), realloc(3)

.SH NOTES
Must define the macro MEMDEBUG before include the memdebug.h header.
This header should be included after the standard headers, particularly
stdlib.h which may be included in some of the other standard headers also.

If the macro MEMDEBUG is not defined
then nothing happens and the standard memory
allocation calls are used.

.SH AUTHOR
R.K.Owen,Ph.D.

.KEY WORDS


NEWS
* Note: Implied Consent -
  If you send in a code patch or fix of one type or another you consent
  to have your name and email address listed here unless you explicitly
  note otherwise.  Whether your name & email is actually listed depends
  on how extensive the change.  (If it's a minor code, document change,
  or spelling correction your name will probably not be listed.)

****************************   Release 3.2 ******************************
Modules 3.2.9
	* Fix the coexistence of load/remove flags (Martin Siegert)
	* Fixed an overlapping string copy (Wilson Snyder & Orion Poplawski)
	* Isolated calls to the new interpreter and get & set env.vars.
	* Optionally use Tcl memory check routines for malloc/realloc.
	* Fixed the "module purge" memory corruption (Poor Yorick)
	* Added support for Ruby (Tammo Tjarks)

Modules 3.2.8
Oct  1, 2010				R.K. Owen (rk@owen.sj.ca.us)
	* Handle the .modulerc/.version in deep modulefile dirs,
	  Patches thanks to Lindsay Todd, and others.
	* Note that .version takes precedence over .modulerc files.
	* Renamed init/python to init/python.py and updated with
	  anonymous patch to eliminate 'sed' references
	* Renamed init/perl to init/perl.pm
	* Added brief section to module.1 on invoking modules for perl & python
	* Added Nathaniel Waisbrot's patch for CMake support with tests.
	* Fixed the MANPATH in modulefiles/modules to use configure generated
	  values.
	* Added the configure --with-man-path so a system specific MANPATH
	  could be used.  Some "man"s radically change their search path
	  when a MANPATH exists and when not.
	  The default is "/usr/man:/usr/share/man" (when both exists).
	* Purging a purged module list does not result in an error.
	* Fixed the configure script for searching for tclConfig.sh
	  and handle --without-x in the tests.

Modules 3.2.7
Jul 30, 2009				R.K. Owen (rk@owen.sj.ca.us)
	* Have init/bash export the module function (thanks to Roy Dragseth)
	* Fixed swap bug when prepending or appending multiple times
	  (fix thanks to Gavin Walker), added tests for same.
	* Have purge remove loaded modules in reverse order
	  (patch contributed by carriees)
	* Added a batch of patches from Gerrit Renker
	  - Goes through each of the two manpages, bringing them up to date.
	  - Fix an error in configure.ac caused by misspelling
	  - Fix typos in help messages and source-code comments
	  - Fix the setup of $MODULESPATH to not have a trailing ':'
	  - Fix segmentation violation occurring when the shelltype is absent
	  - Fix a condition where 'make uninstall' does not cleanse a directory
	  - Fix a build problem which set the executable bit on 'data' files
	  - Cleans up noise in '.modulespath' file
	  - Replaces external sed filter script with internal Makefile.am rule
	  - Adds missing 'zsh' case in etc/global
	  - simplify (t)csh initialisation
	  - Add bash auto-completion for the 'module' command
	  - Adds new 'chdir' modulefile command (sets cwd on module load)
	* versions a & b fixed some bugs

Modules 3.2.6
Oct 29, 2007				R.K. Owen (rk@owen.sj.ca.us)
	* Fixed the locate_module.c:realloc() calls which didn't properly
	  take the element size into account (Fix thanks to Ron Isaacson).
Modules 3.2.5
Feb 14, 2007				R.K. Owen (rk@owen.sj.ca.us)
	* Wrapped all malloc() calls to make sure the size is never 0.
	  This fixes a problem spotted by Francesco Nigro under AIX,
	  and diagnosed by G Lehnert.
	* Updated tests when domainname is unknown.
Modules 3.2.4
Feb  2, 2007				R.K. Owen (rk@owen.sj.ca.us)
	* Fixed bug where "empty" directory paths in the MODULEPATH env.var.
	  needed to be handled gracefully.
Modules 3.2.3
Jun  1, 2006				R.K. Owen (rk@owen.sj.ca.us)
	* Allow version names to be substrings of other version names
	  thanks to dgoldendias at users.sourceforge.net
	* Fixed compiling bug when --enable-log was specified.
	* Made some minor changes to the syslogging and tests so that
	  they all pass.
	* the set-alias command will for Bourne shell derivatives
	  substitute the safe ${1+"$@"} for $*.
	* Added the --delim option to prepend-path and append-path to
	  allow arbitrary delimiters instead of the colon ':'
	* Fixed build problem with regards to the init/* files
	  (solution due to Thomas Zeiser)
Modules 3.2.2
May  4, 2006				R.K. Owen (rk@owen.sj.ca.us)
	* Replaced all strtok() calls with an internal version.
	* Now can append/prepend and empty string '' or colons ':'
	* Added more --with-tclx-* options in case tclxConfig.sh
	  can not be found, but the TclX lib does exist.
	* Expanded the search path for tclConfig.sh.
	* Updated the TclX init call for 8.4
Modules 3.2.1
Feb 13, 2006				R.K. Owen (rk@owen.sj.ca.us)
	* Catch 'exit N' within modulefiles, use this as an exit return code.
	  Has the behavior of loading the modulefile up to the point of exit.
	* For bash, a non-zero exit will append a test that returns false.
	* Added tests for "break" and "exit N"
	* various other bug fixes
Modules 3.2.0
Jan 19, 2006				R.K. Owen (rk@owen.sj.ca.us)
	* default for cache is now "no"
	* Added version_test() to testsuite to make some of the
	  tests conditional on the values generated by --version.
	* Reworked every configure-based file
	  - the configure options help is formatted with AC_HELP_STRING
	  - uses automake to generate most of the Makefile.in
	  - renamed configure.in to configure.ac per autoconf
	  - Versioning is optional now with a configure option
	  - reworked the init scripts for versioning or not
	  - eliminated acconfig.h and put templates in configure.ac
	  - make install now honors the DESTDIR variable for package builds
	  - honors bindir for modulecmd location
	   (but the init scripts are only correct if no versioning)
	* Collected the usage info and now it can be displayed with
	  the --help or -H option (same as help)
	* --version now gives most of the optional configure parameters
	  for easy debugging of problems
	* Made the output deterministic (sorted) and reworked tests also

lib/memdebug.c
static const char RCSID[]="@(#)$Id: memdebug.c,v 1.1 2002/08/27 21:07:41 rkowen Exp $";
static const char AUTHOR[]="@(#)memdebug 1.0 02/10/95 R.K.Owen,PhD";

/* memory - provides a front-end for the memory allocation routines to
 *	help find memory leaks.  Supports only the ANSI-C routines:
 *		  calloc,   free,   malloc,   realloc
 *	call with the following names:
 *		m_calloc, m_free, m_malloc, m_realloc
 *	the output goes to stderr.
 */
/** ** Copyright *********************************************************** **
 ** 									     **
 ** Copyright 1997 by R.K.Owen,Ph.D.		                      	     **
 ** last known email: librko@kooz.sj.ca.us				     **
 **                   rk@owen.sj.ca.us					     **
 ** see LICENSE.LGPL, which must be provided, for details		     **
 ** 									     **
 ** ************************************************************************ **/

#include <stdlib.h>
#include <stdio.h>

FILE *m_output;

FILE *memdebug_output(FILE *output) {

	if (!m_output) m_output = stderr;

	if (output)
		m_output = output;

	return m_output;
}

void *m_calloc(size_t nelem, size_t size, char *file, int line) {
	void *ptr;
	char *null = "(null)";

	if (!m_output) m_output = stderr;

	ptr = calloc(nelem, size);
	if (ptr != NULL)
		(void) fprintf(m_output,"RKOMEM:         calloc : %s %d :"
			" %p %d %d\n",
			file, line, ptr, nelem, size);
	else
		(void) fprintf(m_output,"RKOMEM:         calloc : %s %d :"
			" %s %d %d\n",
			file, line, null, nelem, size);
	return ptr;
}

void m_free(void *ptr, char *file, int line) {
	char *null = "(null)";

	if (!m_output) m_output = stderr;

	if (ptr != NULL)
		(void) fprintf(m_output,"RKOMEM:         free   : %s %d : %p\n",
			file, line, ptr);
	else
		(void) fprintf(m_output,"RKOMEM:         free   : %s %d : %s\n",
			file, line, null);
	free(ptr);
}

void *m_malloc(size_t size, char *file, int line) {
	void *ptr;
	char *null = "(null)";

	if (!m_output) m_output = stderr;

	ptr = malloc(size);
	if (ptr != NULL)
		(void) fprintf(m_output,"RKOMEM:         malloc : %s %d :"
			" %p %d\n",
			file, line, ptr, size);
	else
		(void)fprintf(m_output,"RKOMEM:         malloc : %s %d :"
		" %s %d\n",
			file, line, null, size);
	return ptr;
}

void *m_realloc(void *rptr, size_t size, char *file, int line) {
	void *ptr;
	char *null = "(null)";

	if (!m_output) m_output = stderr;

	ptr = realloc(rptr, size);
	if (rptr != NULL)
		(void) fprintf(m_output,"RKOMEM: realloc_free   : %s %d : %p\n",
		file, line, rptr);

	if (size) {
		(void) fprintf(m_output,"RKOMEM: realloc_malloc : %s %d :",
			file, line);
		if (ptr != NULL)
			(void) fprintf(m_output," %p %d\n", ptr, size);
		else
			(void) fprintf(m_output," %s %d\n", null, size);
	}
	return ptr;
}
